File: Mutual_Exclusion.c
#include <stdio.h>
#include <conio.h>
#include <dos.h>
#include <time.h>

void main() 
{
    int cs = 0, pro = 0;
    double run = 5.0;
    char key = 'a';
    time_t t1, t2;

    clrscr();
    printf("Press a key (except q) to enter a process: ");
    printf("\nPress q at any time to exit\n");

    t1 = time(NULL) - 5;  

    while (key != 'q') 
    {
        if (kbhit()) 
        {
            key = getch();
            if (key != 'q') 
            {
                if (cs != 0) 
                {
                    t2 = time(NULL);
                    if (difftime(t2, t1) > run) 
                    {
                        printf("Process %d ", pro - 1);
                        printf("exited critical section.\n");
                        cs = 0;
                    }
                }
                
                if (cs == 0) 
                {
                    printf("Process %d ", pro);
                    printf("entered critical section.\n");
                    cs = 1;
                    pro++;
                    t1 = time(NULL);
                } 
                else 
                {
                    printf("Error: Another process is currently executing!\n");
                }
            }
        }
    }
}

File: Mutual_Exclusion.py
import time

cs = False
pro = 0  
run = 5.0  
t1 = time.time() - run  

while True:
    key = input("Press a key (except 'q') to enter a process: ")
    if key == 'q':
        print("Exiting..")
        break

    if cs:
        t2 = time.time()
        if (t2-t1) > run:
            print(f"Process {pro - 1} exited critical section.")
            cs = False
        else:
            print("Error: Another process is currently executing!")
    else:
        print(f"Process {pro} entered critical section")
        cs = True
        pro += 1
        t1 = time.time()
        